feat!: Logging refactor & add LoggerPluginInterface

cademirch opened the pull request:
Followup to #2830.

This PR attempts to make Snakemake's logging more modular to make room for a potential logging plugin interface, or other logging/monitoring additions, as well as make logging in Snakemake more standardized.

Currently, most of logging logic is deep within `logging.Logger`, mainly in `text_handler()` with direct calls to `logger.info/warn/etc`. In this PR, this logic has been refactored to more comply with the standard `logging` practices. To do so, three new classes were added:
1. `ColorizingTextHandler` manages adding color to log records and emitting to stream.
2. `DefaultFormatter` formats log records to be printed. Much of the logic from `text_handler()` is here now.
3. `DefaultFilter` filters log records from being emitted based on level and quietness settings.

Together, these classes should output identical logs to the current logging implementation. I've tested this briefly, but further testing would be good.

With these changes, it is much easier to add new `logHandlers` via `logger.addHandler`. New handlers can decide how to format and filter logs themselves, an important characteristic missing in the current implementation that motivated this refactoring (https://github.com/snakemake/snakemake/pull/2830#issuecomment-2092731491). 

This would be a breaking change for `--log-handler-script`, as user provided log handlers will need to subclass `logging.logHandler`. Furthermore, logic for adding the user handler (and formatter/filter) will need to be added.

As for the logging plug interface (https://github.com/snakemake/snakemake/pull/2830#issuecomment-2091510808), I'm not sure that is the route to take, but would appreciate thoughts on this. First, this functionality is already largely provided by `--log-handler-script` and adding a logging plugin would be redundant. Second, I think that development of such plugins would likely be minimal, and coordinated effort in this repo to maintain a good set of logging options would be preferable. For example:
- A standard HTTP log handler send emit log records to a user supplied address on which a 3rd party monitoring app is listening. 
- A handler/formatter that uses [rich](https://github.com/Textualize/rich) to make a prettier/interactive console on which logs are written to. 

I think both of these additions would benefit being standardized to Snakemake, as well as default options to users. 

Additionally I think it would be worthwhile to unify logging in Snakemake. Currently, the use of the wrapper `logging.Logger` class and `setup_logger()` function being called in multiple places makes it difficult to work with the logger and is likely the source of some issues (#3044, #2974, #2882, #2797, #2727). Unifying the logger and its configuration as recommended by the [logging docs](https://docs.python.org/3/howto/logging-cookbook.html#a-cli-application-starter-template) would make it easier to work the the logger. Another option would be to use something like [loguru](https://loguru.readthedocs.io/en/stable/index.html). Whichever option, I think simplifying and unifying the logger should be the next step of this PR. 

One main question I have is how much we would like to preserve the current stderr output formatting. Most of logging logic is in formatting rule info and such. Moving forward we can now try different formatting. Personally, I find the current detailed rule output helpful for sanity checks when developing and doing a dry run. However, when actually running the workflow (especially a large one) the terminal is flooded and errors can quickly be lost which requires searching the log file. Of course this can be alleviated with --quiet. 

Another question is regarding the two calls of setup_logger() brought up in #2797. I still don't understand why these are necessary and understanding this would be helpful in unifying this.

I'm happy to keep pushing this forward but would appreciate your thoughts before continuing @johanneskoester! I


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new CLI option (--logger) for configuring custom loggers.
	- Enhanced structured logging delivers clearer runtime and resource insights, including rule graph logging.
	- Added a new method for logging rule graph information.
	- Added a new enumeration value `REASON` to the `Quietness` class.

- **Refactor**
	- Overhauled the logging architecture across execution, scheduling, and workflow components for improved traceability and error handling.

- **Chores**
	- Updated dependencies and setup configurations to include new logger plugins.
	- Adjusted test configurations to capture all output, ensuring more comprehensive test diagnostics.

- **Bug Fixes**
	- Resolved minor command and formatting inconsistencies for smoother operations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

cademirch commented on snakemake/logging.py:
```diff
@@ -346,42 +781,66 @@
+        custom_level = msg.get("level", "info").lower()
+
+        # Map custom levels to standard Python logging levels
+        log_level = {
+            "debug": _logging.DEBUG,
+            "info": _logging.WARNING,
+            "warning": _logging.WARNING,
+            "error": _logging.ERROR,
+            "critical": _logging.CRITICAL,
+            "run_info": _logging.WARNING,
+            "job_info": _logging.INFO,
+            "group_info": _logging.INFO,
+            "job_error": _logging.ERROR,
+            "group_error": _logging.ERROR,
+            "progress": _logging.INFO,
+            "job_finished": _logging.INFO,
+            "shellcmd": _logging.INFO,
+            "rule_info": _logging.INFO,
+            "dag_debug": (
+                _logging.WARNING if self.debug_dag else _logging.DEBUG
+            ),  # bump debug_dag to warning if we want it
+            "resources_info": _logging.WARNING,
+            "host": _logging.INFO,
+            "job_stats": _logging.WARNING,
+        }.get(
+            custom_level, _logging.INFO
+        )  # Default to INFO if not recognized
+
```
This was intentional to preserve the existing log output coloring:

https://github.com/snakemake/snakemake/blob/19a2ab76b475c5d7728624618bec75a7f1411e8b/snakemake/logging.py#L624-L625

st-- replied:
I agree with the coderabbit bot - Snakemake should not warn on infos, as already pointed out in https://github.com/snakemake/snakemake/issues/2727 - upgrading all infos to warnings makes it impossible to then later filter out the infos without losing all actual warnings. If info level messages should be coloured differently, that ought to be fixed elsewhere, not by conflating it with warnings.

st-- replied:
(Belatedly realized this is on outdated changes, so probably the whole conversation can just get resolved. Sorry for the noise!)

cademirch replied:
I agree, however this mapping is now gone and we're just using direct calls to info, warning etc. Happy to accept changes to specific logging calls that you think should go to different levels

johanneskoester commented:
> As for the logging plug interface (https://github.com/snakemake/snakemake/pull/2830#issuecomment-2091510808), I'm not sure that is the route to take, but would appreciate thoughts on this. First, this functionality is already largely provided by --log-handler-script and adding a logging plugin would be redundant. Second, I think that development of such plugins would likely be minimal, and coordinated effort in this repo to maintain a good set of logging options would be preferable. For example:
>    A standard HTTP log handler send emit log records to a user supplied address on which a 3rd party monitoring app is listening.
>    A handler/formatter that uses [rich](https://github.com/Textualize/rich) to make a prettier/interactive console on which logs are written to.
> I think both of these additions would benefit being standardized to Snakemake, as well as default options to users.

I think it is still fine to develop them as plugins. They can be default dependencies or internal plugins (like with the dryrun, local, and touch executor plugins).

johanneskoester commented:
> One main question I have is how much we would like to preserve the current stderr output formatting. Most of logging logic is in formatting rule info and such. Moving forward we can now try different formatting. Personally, I find the current detailed rule output helpful for sanity checks when developing and doing a dry run. However, when actually running the workflow (especially a large one) the terminal is flooded and errors can quickly be lost which requires searching the log file. Of course this can be alleviated with --quiet.

Sure, the output can be improved in any way. I have no strong feelings about how things are currently logged.

johanneskoester commented:
> Another question is regarding the two calls of setup_logger() brought up in https://github.com/snakemake/snakemake/issues/2797. I still don't understand why these are necessary and understanding this would be helpful in unifying this.

Answered here: https://github.com/snakemake/snakemake/issues/2797#issuecomment-2397668227
Further, one general problem I have with the Snakemake logger is that it uses a global object which kind of has a state. We can keep it like that but one could keep in mind that maybe at some point this should be somehow different and the logger instance should rather be stored in the API call or so. Would be the start of making Snakemake thread safe.

johanneskoester commented on snakemake/logging.py:
```diff
@@ -305,12 +243,504 @@ def log_handler(self, msg):
+        level = msg.get("level", "INFO")  # Default to "INFO" if level not in message
+
+        # Call specific handlers based on the log level
+        if level == "info":
+            return self.handle_info(msg)
+        elif level == "host":
+            return self.handle_host(msg)
+        elif level == "job_info":
+            return self.handle_job_info(msg)
+        elif level == "group_info":
+            return self.handle_group_info(msg)
+        elif level == "job_error":
+            return self.handle_job_error(msg)
+        elif level == "group_error":
+            return self.handle_group_error(msg)
+        elif level == "progress":
+            return self.handle_progress(msg)
+        elif level == "job_finished":
+            return self.handle_job_finished(msg)
+        elif level == "shellcmd":
+            return self.handle_shellcmd(msg)
+        elif level == "rule_info":
+            return self.handle_rule_info(msg)
+        elif level == "d3dag":
+            return self.handle_d3dag(msg)
+        elif level == "dag_debug":
+            return self.handle_dag_debug(msg)
+        elif level == "run_info":
+            return self.handle_run_info(msg)
+        else:
+            return msg["msg"]
```
I guess this is something for Python's new match statement.

cademirch replied:
Good idea, agreed.

johanneskoester commented on snakemake/logging.py:
```diff
@@ -305,12 +243,504 @@ def log_handler(self, msg):
+    """Helper method to show logs."""
+    for f in logs:
+        try:
+            content = open(f, "r").read()
```
I always thought the idiomatic version of this uses a with statement, even if one reads the entire file into memory.

johanneskoester reviewed (changes requested):
See my answers on the PR message content above. Here comes the code review. I like the structured approach to splitting up the code a lot!

cademirch commented:
Thanks for the review and answering questions @johanneskoester. This helps with the direction to take this. Currently working on consolidating the custom log levels. Will try to have more updates by end of the week.

Re logger state: I agree this is a bit tricky, I will try to keep it in mind.

cademirch commented:
> > Another question is regarding the two calls of setup_logger() brought up in #2797. I still don't understand why these are necessary and understanding this would be helpful in unifying this.
> 
> Answered here: [#2797 (comment)](https://github.com/snakemake/snakemake/issues/2797#issuecomment-2397668227) Further, one general problem I have with the Snakemake logger is that it uses a global object which kind of has a state. We can keep it like that but one could keep in mind that maybe at some point this should be somehow different and the logger instance should rather be stored in the API call or so. Would be the start of making Snakemake thread safe.

I've been thinking about this lately as I've been digging into the execmodes and its effect on logging. I'm still a bit confused with the API and have been looking into the LocalExecutor to try and understand, but I can't seem to understand if/where the API is constructed by the executor? Do you have any resources/diagrams of the how this works?

If I understand what you're saying here is that a logger instance could be passed to the constructor of the API instance?  

I'm also not sure how much this matters for local execution to be honest. Currently things are setup such that loggers in subprocess don't log anything (which makes sense to me):

https://github.com/snakemake/snakemake/blob/1874fa86b1ebf52ece6646f32d4a3ed025499183/snakemake/executors/local.py#L84-L85


I was thinking of making this more explicit by just attaching [NullHandler](https://docs.python.org/3/library/logging.handlers.html#logging.NullHandler) to the logger  instead of the StreamHandler when in subproccess mode.


Overall, my understanding is that we only care about logs that come from the main Snakemake process (and by extension any exectuor plugins - which should be in the same process?), so having a global logger makes sense in this way. However its possible I'm missing what you're saying here.

In a remote context, it makes sense that the logger writes to stderr/file, but that doesn't contend with the main Snakemake process so this is okay. In this scenario, we only "care" about the logs from the main Snakemake process, because remote process is totally independent.

cademirch commented:
@johanneskoester some more work here ready for your review.

from the commit message:

> The logger wrapper class (seemingly) largely existed to enable the custom log methods (job_info, progress, etc). However this was quite confusing, especially the usage of the logRecord msg as a dictionary to pass along conextual information.
> 
>  In this commit, the logger class is no longer used. Direct calls to log.info/warn/debug use the `extra` argument to pass a dict of contextual information (jobid, etc). Then the default log formatter/filter can use this info to format and filter logs respectively. This approach provides a much more explicit route of log records and allows potential log plugins to handle logs however they see fit.
> 
> I've put some small functions to setup the logger and such, however this is largely to get things running and I will put more effort into shoring this up if we think this is a good approach.

cademirch commented:
@johanneskoester, this is ready for review! 
## Removal of Logger wrapper class
(From above) In this commit, the logger class is no longer used. Direct calls to log.info/warn/debug use the extra argument to pass a dict of contextual information (jobid, etc). Then the default log formatter/filter can use this info to format and filter logs respectively. This approach provides a much more explicit route of log records and allows potential log plugins to handle logs however they see fit.
## Logger Plugins
I've added support for logger plugins. Logger plugins are relatively simple, all they must do instantiate a class that subclasses `logging.handler`. 
 - [Interface repo](https://github.com/cademirch/snakemake-interface-logger-plugins)
 - [Example plugin repo (uses rich)](https://github.com/cademirch/snakemake-logger-plugin-rich)
### Rich Logger Demo
Here is a gif of the rich logger in action. (Still very much a WIP)
![t-rec_1](https://github.com/user-attachments/assets/1ae0c0b6-3ae4-4005-92ec-52017f1093fe)



## Logger Manager
I created a class to hold most of the logger configuration/setup/destruction logic, `snakemake.logging.LoggerManager`.


## Todo/To discuss (loosely in order of importance)
### Deprecating `--log-handler-script`
This currently wont work with the given changes. I could probably make this work if the `log-handler-script` points to a file that defines a logging.Handler (or subclass), but at that point you almost have a plugin.
### More testing
Will likely need help here. I don't think I've broken anything (except for `--log-handler-script`, `wmslogger`, and `slacklogger`). There may be issues with executor plugins, if they were calling logger.info without a message. 
### Unintuitive quietness and verbosity
Pointed out in #2882 , and I agree. There are 3 ways to control logging verbosity: `quiet`, `debug-dag`, and `verbose`. I think we should try to unify these and make it more clear/more intuitive for a user and developer to control log verbosity. Will start thinking about this now
### Empty log messages
With switching to `logging.Logger` all log calls need a message. Some old log calls didn't have a message, so I put empty strings but we should probably do better.
 
There is probably more, but this is all I can think of now! Let me know if you have any questions.

cademirch commented on snakemake/logging.py:
```diff
@@ -453,324 +612,316 @@ def is_quiet_about(self, msg_type: str):
+            self.add_stream_handler(logging.NullHandler())
+        else:
+            if handler:
+                for h in handler:
```
need to add method for adding file/generic handler plugins

johanneskoester commented:
Fantastic! My thoughts:
## Deprecating --log-handler-script
Deprecating log-handler-script should be fine. Can be a breaking change then (we will get a version bump to Snakemake 9, but the change is small and should affect only very few people).
## Empty log messages
let's get rid of those empty messages entirely while doing the transition. I don't think there are any in executors, but should be easy to find them.

johanneskoester commented:
## Unintuitive quietness and verbosity
Yes, any improvements are  welcome.

cademirch commented:
@johanneskoester Sounds good. One more thing, regarding built in logger plugins, how should that be handled? I was thinking of adding a HTTP and JSON logger as builtin and would prefer to not have those as seperate repos if possible. Could just put them in a loggers subdir like executors? ie: `snakemake/loggers/http.py` and `snakemake/loggers/json.py`

cademirch commented:
Hi @johanneskoester, this is ready for review again! I think this is ready to merge. I've cleaned up a couple things and addressed some of the issues linked above. Namely, I've added a queue handler in front of all logging handlers so that no handlers will block the main snakemake process. I've also added a check to see if the the logger has already been setup to prevent double logging. Finally, added some more logging info, such as printing job errors after the workflow ends (#3214) and more info when a job finishes (#3178).

I still need to work on the rich logging plugin, but that can happen separate from this PR.  I also have plans for a HTTP logger, as well as a browser monitoring webapp (like panoptes) - but those will likely come sometime early next year. I will prioritize the rich logger first. 

Looking forward to your review!

cademirch commented:
bump @johanneskoester

johanneskoester commented on snakemake/cli.py:
```diff
@@ -1851,41 +1846,16 @@ def parse_quietness(quietness) -> Set[Quietness]:
-        wms_logger = logging.WMSLogger(
-            args.wms_monitor, args.wms_monitor_arg, metadata=metadata
-        )
-        log_handler.append(wms_logger.log_handler)
+    for logger_arg in args.logger:
+        plugin = LoggerPluginRegistry().get_plugin(plugin_name=logger_arg)
+        plugin_settings = plugin.get_settings(args)
+        # logger plugin here is the subclass of LoggerPluginBase, we get handler from it in configure_logger.
+        logger_plugin = plugin.logger_plugin(plugin_settings)
+        log_handlers.append(logger_plugin)
 
-    return log_handler
+    return log_handlers
```
No, such error messages are already created inside the plugin framework

johanneskoester commented on snakemake/logging.py:
```diff
@@ -22,428 +26,491 @@ def get_default_exec_mode():
+    """Helper method to show logs."""
+    for f in logs:
+        try:
+            content = open(f, "r").read()
+        except FileNotFoundError:
```
This is a good suggestion.

johanneskoester replied:
Please provide a commitable comment.

johanneskoester replied:
@coderabbitai please provide a commitable suggestion.

cademirch commented:
> > @johanneskoester Sounds good. One more thing, regarding built in logger plugins, how should that be handled? I was thinking of adding a HTTP and JSON logger as builtin and would prefer to not have those as seperate repos if possible. Could just put them in a loggers subdir like executors? ie: `snakemake/loggers/http.py` and `snakemake/loggers/json.py`
> 
> 
> 
> Mhm, I think that having them builtin creates more confusion than having them as real plugins. Will also make it easier to document them in the same way as other loggers via the plugin catalog. The reason for having dryrun, local and touch as builtin executors is rather that they are fundamentally different and more tightly integrated into Snakemake.

Agreed

cademirch commented:
@johanneskoester let me know what you would like to see before merging this. I had put this on the backburner last month, but will have time to work on this now.

johanneskoester commented on snakemake/cli.py:
```diff
@@ -1513,21 +1515,13 @@ def get_argument_parser(profiles=None):
+        nargs="+",
+        default=[],
+        choices=LoggerPluginRegistry().plugins.keys(),
+        help="Specify a custom logger, available via a logger plugin: snakemake_executor_<name>. Multiple can be specified.",
```
```suggestion
        help="Specify one or more custom loggers, available via logger plugins.",
```

johanneskoester commented on snakemake/cli.py:
```diff
@@ -1849,41 +1844,16 @@ def parse_quietness(quietness) -> Set[Quietness]:
+        plugin = LoggerPluginRegistry().get_plugin(plugin_name=logger_arg)
+        plugin_settings = plugin.get_settings(args)
+        # logger plugin here is the subclass of LoggerPluginBase, we get handler from it in configure_logger.
+        logger_plugin = plugin.logger_plugin(plugin_settings)
```
It seems confusing that the logger plugin returns a LoggerPluginBase subclass. Should that be called LogHandlerBase, and the method `plugin.log_handler(plugin_settings)`?

cademirch replied:
I was confused looking at this again, so agreed. Just so we're on the same page you're suggesting changing this in: `src/snakemake_interface_logger_plugins/registry/plugin.py`


```diff
 @dataclass
 class Plugin(PluginBase):
-    logger_plugin: object
+    log_handler: object
     _logger_settings_cls: Optional[Type[LoggerPluginSettingsBase]]
     _name: str
```

And then in `src/snakemake_interface_logger_plugins/base.py`

```diff
-class LoggerPluginBase(ABC):
+class LogHandlerBase(ABC):
     def __init__(
         self,
         settings: Optional[LoggerPluginSettingsBase],
     ) -> None:
         self.settings = settings
         self.__post__init()
```

johanneskoester replied:
Yes. In addition, I would use `log_handler: LogHandlerBase` as type annotation in `src/snakemake_interface_logger_plugins/registry/plugin.py`.

cademirch replied:
Resolved in https://github.com/cademirch/snakemake-interface-logger-plugins/commit/2b23736a53a569d2cf9fc702d4eed39d283c39bd

johanneskoester commented on snakemake/cli.py:
```diff
@@ -1849,41 +1844,16 @@ def parse_quietness(quietness) -> Set[Quietness]:
 
 
 def setup_log_handlers(args, parser):
```
Since we anyway produce a "hidden" breaking change in the API because the objects passed to snakemake.settings.types.OutputSettings.log_handlers are now coming from a different base class, we should probably also change the type there overall and expect a list of strings with logger plugin names in that settings class, while moving the collection of the corresponding plugins into the API classes, analogously to storage and exectuor plugins.

cademirch replied:
Not sure I follow here. Is this what you're thinking:
```diff
@dataclass
class OutputSettings(SettingsBase):
    printshellcmds: bool = False
    nocolor: bool = False
    quiet: Optional[AnySet[Quietness]] = None
    debug_dag: bool = False
    verbose: bool = False
    show_failed_logs: bool = False
-   log_handlers: Sequence[object] = tuple()
+   log_handlers: Sequence[str] = tuple()
    keep_logger: bool = False
    stdout: bool = False
    benchmark_extended: bool = False
```

johanneskoester replied:
Yes, and the strings are just the names. Then, the instantiation can happen inside of the API classes and the users of the API don't not need to instantiate those things themselves.

johanneskoester commented on snakemake/logging.py:
```diff
@@ -22,428 +26,491 @@ def get_default_exec_mode():
+    """Helper method to show logs."""
+    for f in logs:
+        try:
+            content = open(f, "r").read()
```
```suggestion
            with open(f, "r") as log_file:
                content = log_file.read()
```

johanneskoester commented on snakemake/logging.py:
```diff
@@ -22,428 +26,491 @@ def get_default_exec_mode():
+            yield logfile_header
+            return
+        yield logfile_header
+        max_len = min(max(max(len(l) for l in lines), len(logfile_header)), 80)
```
```suggestion
        max_len = min(max(max(len(line) for line in lines), len(logfile_header)), 80)
```

johanneskoester reviewed (changes requested):
Fantastic and very clean work! A few things below.

cademirch commented on snakemake/logging.py:
```diff
@@ -3,17 +3,21 @@
+
 import textwrap
+from typing import Optional
+from snakemake_interface_logger_plugins.base import LoggerPluginBase
```
This depedency is in the test-env yaml

johanneskoester replied:
But it should also be in setup.cfg shouldn't it?

cademirch replied:
You're right, thanks!

johanneskoester commented on snakemake/logging.py:
```diff
@@ -22,428 +26,493 @@ def get_default_exec_mode():
+                self.stream.write(self.decorate(record, formatted_message))
+                self.stream.write(getattr(self, "terminator", "\n"))
+                self.flush()
+            except BrokenPipeError as e:
+                raise e
+            except (KeyboardInterrupt, SystemExit):
+                pass  # Ignore exceptions for these cases
+            except Exception as e:
+                self.handleError(record)
```
```suggestion
            except BrokenPipeError:
                raise
            except (KeyboardInterrupt, SystemExit):
                pass  # Ignore exceptions for these cases
            except Exception as e:
                self.handleError(record)
```

cademirch commented:
> **Actionable comments posted: 0**
> 
> 🧹 Nitpick comments (2)
> > snakemake/logging.py (2)> `424-433`: **Simplify filter logic for better readability.**
> > > The filter logic can be simplified without changing functionality.
> > > Apply these simplifications:
> > > ```diff
> > > -        if level in quietness_map:
> > > -            if self.is_quiet_about(quietness_map[level]):
> > > -                return False
> > > +        if level in quietness_map and self.is_quiet_about(quietness_map[level]):
> > > +            return False
> > > 
> > > -        if level == "dag_debug" and not self.debug_dag:
> > > -            return False
> > > -        return True
> > > +        return not (level == "dag_debug" and not self.debug_dag)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 🧰 Tools
> > > 
> > > 🪛 Ruff (0.8.2)
> > > 425-426: Use a single `if` statement instead of nested `if` statements
> > > Combine `if` statements using `and`
> > > (SIM102)
> > > 
> > > 430-433: Return the condition `not (level == "dag_debug" and not self.debug_dag)` directly
> > > Replace with `return not (level == "dag_debug" and not self.debug_dag)`
> > > (SIM103)
> > > 
> > > 
> > > 
> > > `529-534`: **Remove unused exception variable.**
> > > The exception variable `e` in the final except block is not used.
> > > Apply this diff:
> > > ```diff
> > >             except BrokenPipeError:
> > >                 raise
> > >             except (KeyboardInterrupt, SystemExit):
> > >                 pass  # Ignore exceptions for these cases, all errors have been handled before.
> > > -            except Exception as e:
> > > +            except Exception:
> > >                 self.handleError(record)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > 🧰 Tools
> > > 
> > > 🪛 Ruff (0.8.2)
> > > 533-533: Local variable `e` is assigned to but never used
> > > Remove assignment to unused variable `e`
> > > (F841)
> 
> 📜 Review details

The first suggestion is less readable imo. I prefer explicit return statements ie `return True` not `return condition`

johanneskoester commented:
> > **Actionable comments posted: 0**
> > 🧹 Nitpick comments (2)
> > > snakemake/logging.py (2)> `424-433`: **Simplify filter logic for better readability.**
> > > > The filter logic can be simplified without changing functionality.
> > > > Apply these simplifications:
> > > > ```diff
> > > > -        if level in quietness_map:
> > > > -            if self.is_quiet_about(quietness_map[level]):
> > > > -                return False
> > > > +        if level in quietness_map and self.is_quiet_about(quietness_map[level]):
> > > > +            return False
> > > > 
> > > > -        if level == "dag_debug" and not self.debug_dag:
> > > > -            return False
> > > > -        return True
> > > > +        return not (level == "dag_debug" and not self.debug_dag)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 🧰 Tools
> > > > 🪛 Ruff (0.8.2)
> > > > 425-426: Use a single `if` statement instead of nested `if` statements
> > > > Combine `if` statements using `and`
> > > > (SIM102)
> > > > 430-433: Return the condition `not (level == "dag_debug" and not self.debug_dag)` directly
> > > > Replace with `return not (level == "dag_debug" and not self.debug_dag)`
> > > > (SIM103)
> > > > `529-534`: **Remove unused exception variable.**
> > > > The exception variable `e` in the final except block is not used.
> > > > Apply this diff:
> > > > ```diff
> > > >             except BrokenPipeError:
> > > >                 raise
> > > >             except (KeyboardInterrupt, SystemExit):
> > > >                 pass  # Ignore exceptions for these cases, all errors have been handled before.
> > > > -            except Exception as e:
> > > > +            except Exception:
> > > >                 self.handleError(record)
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 🧰 Tools
> > > > 🪛 Ruff (0.8.2)
> > > > 533-533: Local variable `e` is assigned to but never used
> > > > Remove assignment to unused variable `e`
> > > > (F841)
> > 
> > 
> > 📜 Review details
> 
> The first suggestion is less readable imo. I prefer explicit return statements ie `return True` not `return condition`

ok, fine with that, matter of style.

cademirch commented:
> > > **Actionable comments posted: 0**
> > > 🧹 Nitpick comments (2)
> > > > snakemake/logging.py (2)> `424-433`: **Simplify filter logic for better readability.**
> > > > > The filter logic can be simplified without changing functionality.
> > > > > Apply these simplifications:
> > > > > ```diff
> > > > > -        if level in quietness_map:
> > > > > -            if self.is_quiet_about(quietness_map[level]):
> > > > > -                return False
> > > > > +        if level in quietness_map and self.is_quiet_about(quietness_map[level]):
> > > > > +            return False
> > > > > 
> > > > > -        if level == "dag_debug" and not self.debug_dag:
> > > > > -            return False
> > > > > -        return True
> > > > > +        return not (level == "dag_debug" and not self.debug_dag)
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > 🧰 Tools
> > > > > 🪛 Ruff (0.8.2)
> > > > > 425-426: Use a single `if` statement instead of nested `if` statements
> > > > > Combine `if` statements using `and`
> > > > > (SIM102)
> > > > > 430-433: Return the condition `not (level == "dag_debug" and not self.debug_dag)` directly
> > > > > Replace with `return not (level == "dag_debug" and not self.debug_dag)`
> > > > > (SIM103)
> > > > > `529-534`: **Remove unused exception variable.**
> > > > > The exception variable `e` in the final except block is not used.
> > > > > Apply this diff:
> > > > > ```diff
> > > > >             except BrokenPipeError:
> > > > >                 raise
> > > > >             except (KeyboardInterrupt, SystemExit):
> > > > >                 pass  # Ignore exceptions for these cases, all errors have been handled before.
> > > > > -            except Exception as e:
> > > > > +            except Exception:
> > > > >                 self.handleError(record)
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > 🧰 Tools
> > > > > 🪛 Ruff (0.8.2)
> > > > > 533-533: Local variable `e` is assigned to but never used
> > > > > Remove assignment to unused variable `e`
> > > > > (F841)
> > > 
> > > 
> > > 📜 Review details
> > 
> > 
> > The first suggestion is less readable imo. I prefer explicit return statements ie `return True` not `return condition`
> 
> ok, fine with that, matter of style.

Great. Lmk what else we should do here. Happy to look into #1902 before or after merging this

cademirch commented:
I've added some more to this PR. 

1. `--quiet` now accepts `reason` which will suppress reason info in the logs.
2. The `LoggerManager` class has been cleaned up significantly and simplified
3. The logger is only setup exactly once, either in `snakemake.cli` or `snakemake.api`
4. `snakemake.settings.types.OutputSettings` now inherits from `snakemake_interface_logger_plugins.settings.LoggerSettingsInterface` such that LogPlugins have easy access to user defined log variables such as quietness, verbosity, etc.

@johanneskoester I think we are getting closer on this! Please let me know what you think.

cademirch commented:
Hi @johanneskoester - Unfortunately I can't seem to exactly replicate the environment that the actions run in, which has made troubleshooting a bit of pain, however, I've got some more of the tests passing.I caught some overlooked old logger custom method calls, causing issues with shell() in run block. I also removed the setting of `printshellcmd` from the script preamble, not sure what its purpose was there, happy to come up with a way to refactor instead of removing.


Currently the first test to fail on the actions is https://github.com/cademirch/snakemake/blob/logging-refactor/tests/tests.py#L1096, and based on the error, the issue seems more related to the test itself? I'm not very familiar with the test suite, and this one confuses me! Any thoughts? 


Would appreciate any help and advice for running the test suite.

cademirch commented:
@johanneskoester all tests pass, would appreciate your review. There is still room for improvement, particularly in where and what we log (handful of empty log messages, could use logging on workflow start, etc.) But I think those can come later.

johanneskoester commented on snakemake/script/__init__.py:
```diff
@@ -30,7 +30,7 @@
 from snakemake.deployment import singularity
 from snakemake.exceptions import WorkflowError
-from snakemake.logging import logger
+from snakemake.logging import logger_manager, logger
```
@cademirch is the import needed?

johanneskoester commented on snakemake/workflow.py:
```diff
@@ -1296,7 +1317,11 @@ def execute(
                     if len(self.dag):
-                        logger.run_info("\n".join(self.dag.stats()))
+                        stats_msg, stats_dict = self.dag.stats()
+                        logger.warning(
```
logger.info here, correct?

cademirch replied:
Warning was correct to coerce it yellow, but coloring based on the log type will resolve this. I will push some changes soon to address this.

cademirch replied:
resolved w/ LogEvent enum

cademirch commented:
Pushed some more changes here. 

I added an enum `LogEvent` to replace the `level` field  in the extra dict and renamed it to `event`. I chose "event" because it better represents what we're tracking and avoids confusion with the standard log severity levels (INFO, DEBUG, ERROR, etc.) 

LogEvent could (perhaps should?) live in logger-interface repo so that plugins can have access to this without making an interface class, but I will defer to you on this @johanneskoester. 

Using the LogEvent in the DefaultFormatter and Handler I've tried to make the log coloring match how it currently is in main. It's a bit difficult since in main, `logger.info` actually calls `logger.warn`:

https://github.com/snakemake/snakemake/blob/04f89d330dd94baa51f41bc796392f85bccbd231/snakemake/logging.py#L385-L386

https://github.com/snakemake/snakemake/blob/04f89d330dd94baa51f41bc796392f85bccbd231/snakemake/logging.py#L624-L625 

Currently all tests pass, and I've visually inspected log output on a handful of test cases but can't promise the formatting will be 1:1 to main in all scenarios!

johanneskoester commented:
> LogEvent could (perhaps should?) live in logger-interface repo so that plugins can have access to this without making an interface class, but I will defer to you on this @johanneskoester.
Yes, that makes sense. 

> Currently all tests pass, and I've visually inspected log output on a handful of test cases but can't promise the formatting will be 1:1 to main in all scenarios!
Fine, thanks!

So, I would be fine with merging this once the new enum is in the logger interface.
